docker/ubuntu-14.04-cpu-all-options/Dockerfile

```diff
@@ -0,0 +1,72 @@
+FROM ubuntu:14.04
+MAINTAINER Aaron Markham <aaronmarkham@fb.com>
+
+# caffe2 install with cpu support
+
+########## REQUIRED DEPENDENCIES ################
+RUN apt-get update && apt-get install -y --no-install-recommends \
+    python-dev \
+    python-pip \
+    git \
+    build-essential \
+    cmake \
+    libprotobuf-dev \
+    protobuf-compiler \
+    libgoogle-glog-dev
+
+RUN git clone --recursive https://github.com/caffe2/caffe2.git && \
+    pip install --upgrade pip setuptools wheel && \
+    pip install future hypothesis numpy protobuf six
+
+########## OPTIONAL DEPENDENCIES ################
+
+RUN apt-get install -y --no-install-recommends \
+    libgflags-dev \
+    libgoogle-glog-dev \
+    libgflags2 \
+    libgtest-dev \
+    libiomp-dev \
+    libleveldb-dev \
+    liblmdb-dev \
+    libopencv-dev \
+    libopenmpi-dev \
+    libsnappy-dev \
+    openmpi-bin \
+    openmpi-doc \
+    python-numpy \
+    python-pydot \
+    python-setuptools \
+    python-scipy \
+    wget
+
+RUN pip install \
+    flask \
+    graphviz \
+    hypothesis \
+    jupyter \
+    matplotlib \
+    pydot \
+    python-nvd3 \
+    pyyaml \
+    requests \
+    scikit-image \
+    scipy \
+    setuptools \
+    tornado
```

Comment:
I'm going to leave this for the next round since I've already made so many changes. Plus I really think it would be good to explore pinning versions of the dependencies when doing this, and that will take more research and testing. I could pin versions based on what I have as a working build, but it wouldn't really reflect minimum possible versions that would work. Happy to hear of any good approaches to this.